libkbfs/prefetcher.go

```diff
@@ -496,6 +586,19 @@ func (p *blockPrefetcher) run(testSyncCh <-chan struct{}) {
 			p.log.Debug("canceling prefetch for block %s", blockID)
 			// Walk up the block tree and delete every parent.
 			p.applyToParentsRecursive(p.cancelPrefetch, blockID, pre)
+		case reqInt := <-p.prefetchRescheduleCh.Out():
+			req := reqInt.(*prefetchRequest)
+			blockID := req.ptr.ID
+			pre, ok := p.prefetches[blockID]
+			if !ok {
+				pre = p.newPrefetch(1, false, req)
+				p.prefetches[blockID] = pre
+			}
+			p.log.Debug("rescheduling top-block prefetch for block %s", blockID)
```

Comment:
> But I guess that old parent backpointers don't get cleaned up on a cancel

I don't think this is true--`CancelPrefetch` does an `applyToParentsRecursive`.

Reply:
But if you have a block path `a1 -> b`, and then we get an update where `a1 -> a2` but `b` doesn't change, then `CancelPrefetch(a1)` is called.  But that doesn't update the prefetch state for `b` to take away `a1` as a parent pointer, right?

Reply:
@strib correct, it doesn't take away `a1` as a parent pointer, but then as soon as `b`'s prefetch happens, it'll hit this line, which will remove it from its parents:
https://github.com/keybase/kbfs/blob/9291860e2465a194d9916d53cc683caf6243d919/libkbfs/prefetcher.go#L171

Reply:
Which actually brings to mind another point: I wonder if that should now be:
```go
parent, ok := p.prefetches[b]
if !ok {
	parent, ok = p.rescheduled[b]
	if !ok {
		delete(pre.parents, b)
	}
}
```

Reply:
> I wonder if that should now be:

Hmm, I don't think so.  If the prefetch has been legitimately cancel, as in the parent-being-upgraded case, we don't want to bother rescheduling it.  So with your clarification and the behavior of `folderBlockOps` to cancel old prefetches, I think it's very unlikely that a block will actually have multiple parents for any meaningful amount of time.  So it seems to me like everything here should work correctly -- in most cases, it'll reschedule the prefetch of the top-most block it can find in the tree.

Please correct me if I've missed something.

Reply:
I agree that it should be rare.

> If the prefetch has been legitimately cancel, as in the parent-being-upgraded case, we don't want to bother rescheduling it.

100% agree about that case. But now when you reschedule, you move a block out of `p.prefetches` and into `p.rescheduled`. We probably shouldn't lose parent tracking in that case...

Reply:
Then again, maybe we should. Since the top-most block will be doing the prefetching in parent tracking, and we will have lost the full chain.

E.g. `a -> b -> c -> d -> e`. We reschedule `c`, which blasts away `c`, and `b`, and reschedules `a`. Then we prefetch `e`, which reschedules to `d`. Not sure that's the right answer here.

Reply:
Yeah I think the scenario you outline just above is a good-enough one to start with for now.  I don't think we can do much better without trying to keep the `*prefetchRequest` for rescheduled requests around somewhere.  I propose just keeping `applyToParentsRecursive` as it is for now, and revisiting later if needed.  I'll add a comment there though.